stacker/src/memory_arena.rs

```diff
@@ -134,44 +128,66 @@ impl MemoryArena {
     pub fn read<Item: Copy + 'static>(&self, addr: Addr) -> Item {
         load(self.slice(addr, mem::size_of::<Item>()))
     }
+    #[inline]
+    fn get_page(&self, page_id: usize) -> &Page {
+        unsafe { self.pages.get_unchecked(page_id) }
+    }
+    #[inline]
+    fn get_page_mut(&mut self, page_id: usize) -> &mut Page {
+        unsafe { self.pages.get_unchecked_mut(page_id) }
+    }
 
     #[inline]
     pub fn slice(&self, addr: Addr, len: usize) -> &[u8] {
-        self.pages[addr.page_id()].slice(addr.page_local_addr(), len)
+        self.get_page(addr.page_id())
+            .slice(addr.page_local_addr(), len)
     }
 
     #[inline]
     pub fn slice_from(&self, addr: Addr) -> &[u8] {
-        self.pages[addr.page_id()].slice_from(addr.page_local_addr())
+        self.get_page(addr.page_id())
+            .slice_from(addr.page_local_addr())
     }
 
     #[inline]
     pub fn slice_mut(&mut self, addr: Addr, len: usize) -> &mut [u8] {
-        self.pages[addr.page_id()].slice_mut(addr.page_local_addr(), len)
+        self.get_page_mut(addr.page_id())
+            .slice_mut(addr.page_local_addr(), len)
+    }
+
+    #[inline]
+    fn add_page(&mut self, len: usize) -> usize {
+        let new_page_id = self.pages.len();
+        let mut page = Page::new(new_page_id);
+        page.len = len;
+        self.pages.push(page);
+        new_page_id
     }
 
     /// Allocates `len` bytes and returns the allocated address.
+    #[inline]
     pub fn allocate_space(&mut self, len: usize) -> Addr {
         let page_id = self.pages.len() - 1;
-        if let Some(addr) = self.pages[page_id].allocate_space(len) {
+        if let Some(addr) = self.get_page_mut(page_id).allocate_space(len) {
             return addr;
         }
-        self.add_page().allocate_space(len).unwrap()
+        let new_page_id = self.add_page(len);
+        Addr::new(new_page_id, 0)
     }
 }
 
 struct Page {
     page_id: usize,
     len: usize,
-    data: Box<[u8]>,
+    data: Box<[u8; PAGE_SIZE]>,
 }
 
 impl Page {
     fn new(page_id: usize) -> Page {
         Page {
             page_id,
             len: 0,
-            data: vec![0u8; PAGE_SIZE].into_boxed_slice(),
+            data: Box::new([0u8; PAGE_SIZE]),
```

Comment:
Did it help? If it does not I'd rather avoid the stack allocation,

Reply:
Both are heap allocations. It removes the bounds check in combination with the mask, since the size is known to the compiler

Reply:
They both end up on the heap, but Rust will often allocate the array on the stack before moving it to the heap, c.f. https://github.com/rust-lang/rust/issues/53827

Reply:
I don't see that issue looking at the assembly, but `vec![0, PAGE_SIZE]` is more efficient since it calls alloc zeroed vs. alloc + memset

I"ll switch to `vec![0u8; PAGE_SIZE].into_boxed_slice().try_into().unwrap()`